compiler/rustc_middle/src/ty/vtable.rs

```diff
@@ -0,0 +1,93 @@
+use std::convert::TryFrom;
+
+use crate::mir::interpret::{alloc_range, AllocId, Allocation, Pointer, Scalar};
+use crate::ty::{self, DefId, SubstsRef, Ty, TyCtxt};
+
+#[derive(Clone, Copy, Debug, PartialEq, HashStable)]
+pub enum VtblEntry<'tcx> {
+    MetadataDropInPlace,
+    MetadataSize,
+    MetadataAlign,
+    Vacant,
+    Method(DefId, SubstsRef<'tcx>),
+}
+
+pub const COMMON_VTABLE_ENTRIES: &[VtblEntry<'_>] =
+    &[VtblEntry::MetadataDropInPlace, VtblEntry::MetadataSize, VtblEntry::MetadataAlign];
+
+pub const COMMON_VTABLE_ENTRIES_DROPINPLACE: usize = 0;
+pub const COMMON_VTABLE_ENTRIES_SIZE: usize = 1;
+pub const COMMON_VTABLE_ENTRIES_ALIGN: usize = 2;
+
+impl<'tcx> TyCtxt<'tcx> {
+    pub fn vtable_allocation(
+        self,
+        ty: Ty<'tcx>,
+        poly_trait_ref: Option<ty::PolyExistentialTraitRef<'tcx>>,
+    ) -> AllocId {
+        let tcx = self;
+        let vtables_cache = tcx.vtables_cache.lock();
+        if let Some(alloc_id) = vtables_cache.get(&(ty, poly_trait_ref)).cloned() {
+            return alloc_id;
+        }
+        drop(vtables_cache);
+        let param_env = ty::ParamEnv::reveal_all();
```

Comment:
```suggestion

        let param_env = ty::ParamEnv::reveal_all();
```
A new logical section of the code start here, let's indicate that. :)

Also, I learned to be wary of `reveal_all`. Is this truly the right `ParamEnv` here?

Reply:
probably not, but unproblematic in practice for now. It would still be cleaner to pass in the param env as an argument. codegen can then use reveal all, as it does everywhere, and CTFE can use its real param env

Reply:
@oli-obk I have a question here. With different paramenv, does this mean CTFE and codegen cannot reuse the same set of cached allocation?

Reply:
Most of the time they will, the only difference is when we compute a vtable in a generic constant. the vtable computation should be independent of the param env as far as i can tell, but i don't know how to be sure of that. the param env contains the trait bounds on the generic parameters in scope. we may be able to compute a vtable for a generic type, if the generic param is irrelevant for the size, align, drop logic and methods. similarly we may be able to compute the vtable for a generic trait, if the generic param is irrelevant. i don't know if this is reasonable, so i have a better alternative: don't pass a param env, but instead assert that ty, substs (from VtblEntry::Method) and poly_trait_ref are not needs_subst, because then everything should be fine. you are already unwrapping the results of param env'ed calls, so we aren't adding new sources of panics.

Reply:
Done.